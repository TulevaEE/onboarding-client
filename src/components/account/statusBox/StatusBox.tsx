import React from 'react';
import { Message } from 'retranslate';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { StatusBoxRow } from './statusBoxRow/StatusBoxRow';
import SecondPillarStatusBox from './secondPillarStatusBox';
import { Fund, UserConversion } from './types';
import { StatusBoxLoader } from './StatusBoxLoader';
import { StatusBoxTitle } from './StatusBoxTitle';

interface StatusBoxType {
  memberNumber: number | null;
  conversion?: UserConversion;
  loading: boolean;
  secondPillarFunds: Fund[];
  thirdPillarFunds: Fund[];
}

export const StatusBox: React.FunctionComponent<StatusBoxType> = ({
  conversion,
  memberNumber = null,
  loading = false,
  secondPillarFunds = [],
  thirdPillarFunds = [],
}) => {
  if (!conversion || !secondPillarFunds || !thirdPillarFunds) {
    return <StatusBoxLoader />;
  }

  const payTuleva3 = !(
    conversion.thirdPillar.selectionComplete &&
    conversion.thirdPillar.transfersComplete &&
    conversion.thirdPillar.paymentComplete
  );

  const isTulevaMember = memberNumber != null;

  const tulevaData = isTulevaMember
    ? [<Message params={{ memberNumber }}>account.member.statement</Message>]
    : [<Message>account.non.member.statement</Message>];

  const thirdPillarActiveFunds = thirdPillarFunds
    .filter((fund) => fund.activeFund)
    .map(({ name }) => name);

  return (
    <>
      <StatusBoxTitle />

      <div className="card card-secondary">
        <SecondPillarStatusBox />

        <StatusBoxRow
          ok={!payTuleva3}
          showAction={!loading}
          name={<Message>account.status.choice.pillar.third</Message>}
          lines={thirdPillarActiveFunds}
        >
          <Link to="/3rd-pillar-flow" className="btn btn-light">
            <Message>account.status.choice.pay.tuleva.3</Message>
          </Link>
        </StatusBoxRow>

        <StatusBoxRow
          last
          ok={isTulevaMember}
          showAction={!loading}
          name={<Message>account.status.choice.tuleva</Message>}
          lines={tulevaData}
        >
          {!isTulevaMember && (
            <Link to="/join" className="btn btn-light">
              <Message>account.status.choice.join.tuleva</Message>
            </Link>
          )}
        </StatusBoxRow>
      </div>
      <div className="mt-3">
        <small className="text-muted">
          <Message
            params={{
              contribution: (conversion.thirdPillar.contribution.yearToDate || 0).toString(),
            }}
            dangerouslyTranslateInnerHTML="account.status.yearToDateContribution"
          />
        </small>
      </div>
    </>
  );
};

const mapStateToProps = (state: {
  login: {
    userConversion: UserConversion;
    loadingUserConversion: boolean;
    user: { memberNumber: number };
  };
  exchange: { sourceFunds: Fund[] };
  thirdPillar: { sourceFunds: Fund[] };
}) => ({
  memberNumber: (state.login.user || {}).memberNumber,
  conversion: state.login.userConversion,
  loading: state.login.loadingUserConversion,
  secondPillarFunds: state.exchange.sourceFunds,
  thirdPillarFunds: state.thirdPillar.sourceFunds,
});

export default connect(mapStateToProps)(StatusBox);
