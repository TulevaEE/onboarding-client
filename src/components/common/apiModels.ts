export type Application =
  | TransferApplication
  | StopContributionsApplication
  | ResumeContributionsApplication
  | EarlyWithdrawalApplication
  | WithdrawalApplication
  | PaymentApplication;

export type PaymentApplication = BaseApplication<
  ApplicationType.PAYMENT,
  {
    amount: number;
    currency: Currency;
    targetFund: Fund;
  }
>;

export type TransferApplication = BaseApplication<
  ApplicationType.TRANSFER,
  {
    sourceFund: Fund;
    exchanges: {
      targetFund: Fund | null;
      targetPik: string | null;
      amount: number;
    }[];
    cancellationDeadline: string;
  }
>;

export type StopContributionsApplication = BaseApplication<
  ApplicationType.STOP_CONTRIBUTIONS,
  {
    stopTime: string;
    earliestResumeTime: string;
    cancellationDeadline: string;
  }
>;

export type ResumeContributionsApplication = BaseApplication<
  ApplicationType.RESUME_CONTRIBUTIONS,
  {
    resumeTime: string;
    cancellationDeadline: string;
  }
>;

export type EarlyWithdrawalApplication = BaseApplication<
  ApplicationType.EARLY_WITHDRAWAL,
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export type WithdrawalApplication = BaseApplication<
  ApplicationType.WITHDRAWAL,
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export enum ApplicationType {
  TRANSFER = 'TRANSFER',
  STOP_CONTRIBUTIONS = 'STOP_CONTRIBUTIONS',
  RESUME_CONTRIBUTIONS = 'RESUME_CONTRIBUTIONS',
  EARLY_WITHDRAWAL = 'EARLY_WITHDRAWAL',
  WITHDRAWAL = 'WITHDRAWAL',
  PAYMENT = 'PAYMENT',
}

export enum ApplicationStatus {
  PENDING = 'PENDING',
  COMPLETE = 'COMPLETE',
  FAILED = 'FAILED',
}

export interface BaseApplication<Type extends ApplicationType, Details> {
  id: number;
  status: ApplicationStatus;
  creationTime: string;

  type: Type;
  details: Details;
}

export interface Fund {
  isin: string;
  name: string;
  pillar: Pillar;
  managementFeeRate: number;
  ongoingChargesFigure: number;
  fundManager: FundManager;
  status: FundStatus;
}

export type Pillar = 2 | 3;

enum FundStatus {
  ACTIVE = 'ACTIVE',
  LIQUIDATED = 'LIQUIDATED',
  SUSPENDED = 'SUSPENDED',
  CONTRIBUTIONS_FORBIDDEN = 'CONTRIBUTIONS_FORBIDDEN',
  PAYOUTS_FORBIDDEN = 'PAYOUTS_FORBIDDEN',
}

export interface FundBalance {
  fund: Fund;
  value: number;
  unavailableValue: number;
  currency: string;
  activeContributions: boolean;
  contributions: number;
  subtractions: number;
  profit: number;
}

export interface SourceFund {
  fundManager: FundManager;
  activeFund: boolean;
  name: string;
  pillar: Pillar;
  managementFeePercent: string;
  isin: string;
  price: number;
  unavailablePrice: number;
  currency: string;
  ongoingChargesFigure: number;
  contributions: number;
  subtractions: number;
  profit: number;
}

export interface FundManager {
  name: string;
}

export interface CancellationMandate {
  mandateId: number;
}

export interface Mandate {
  id: number;
}

interface Address {
  street: string;
  districtCode: string;
  postalCode: string;
  countryCode: string;
}

export interface User {
  id: number;
  personalCode: string;
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;
  memberNumber: number;
  pensionAccountNumber: string;
  secondPillarPikNumber: string;
  address: Address;
  secondPillarActive: boolean;
  thirdPillarActive: boolean;
  age: number;
}

export interface UserConversion {
  secondPillar: Conversion;
  thirdPillar: Conversion;
}

export interface Amount {
  yearToDate: number;
  total: number;
}

export interface Conversion {
  selectionComplete: boolean;
  transfersComplete: boolean;
  paymentComplete: boolean;
  pendingWithdrawal: boolean;
  contribution: Amount;
  subtraction: Amount;
}

export interface InitialCapital {
  membershipBonus: number;
  capitalPayment: number;
  currency: string;
  unvestedWorkCompensation: number;
  workCompensation: number;
  profit: number;
}

export interface AmlCheck {
  type: string;
  success: boolean;
}

export interface HttpError {
  body: { errors: [{ code: string }] };
}

export interface Token {
  accessToken: string;
}

export interface MobileSignatureResponse {
  challengeCode: string;
}

export interface MobileSignatureStatusResponse {
  statusCode: string;
  challengeCode: string;
}

export interface IdCardSignatureResponse {
  hash: string;
}

export interface IdCardSignatureStatusResponse {
  statusCode: string;
}

export interface Payment {
  amount: number;
  currency: Currency;
  bank: string;
}

export type Currency = 'EUR';
