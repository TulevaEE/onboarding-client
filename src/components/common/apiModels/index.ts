import { PaymentRate } from '../../flows/secondPillarPaymentRate/types';

export type Application =
  | TransferApplication
  | StopContributionsApplication
  | ResumeContributionsApplication
  | EarlyWithdrawalApplication
  | WithdrawalApplication
  | PaymentApplication
  | SavingsFundPaymentApplication
  | PaymentRateApplication
  | FundPensionOpeningApplication
  | PartialWithdrawalApplication
  | ThirdPillarWithdrawalApplication;

export type PaymentRateApplication = BaseApplication<
  'PAYMENT_RATE',
  {
    paymentRate: number;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export type PaymentApplication = BaseApplication<
  'PAYMENT',
  {
    amount: number;
    currency: Currency;
    targetFund: Fund;
  }
>;

export type SavingsFundPaymentApplication = BaseApplication<
  'SAVING_FUND_PAYMENT',
  {
    amount: number;
    currency: Currency;
    cancellationDeadline: string;
    fulfillmentDeadline: string;
    paymentId: string;
  }
>;

export type TransferApplication = BaseApplication<
  'TRANSFER',
  {
    sourceFund: Fund;
    exchanges: {
      targetFund: Fund | null;
      targetPik: string | null;
      amount: number;
    }[];
    cancellationDeadline: string;
  }
>;

export type StopContributionsApplication = BaseApplication<
  'STOP_CONTRIBUTIONS',
  {
    stopTime: string;
    earliestResumeTime: string;
    cancellationDeadline: string;
  }
>;

export type ResumeContributionsApplication = BaseApplication<
  'RESUME_CONTRIBUTIONS',
  {
    resumeTime: string;
    cancellationDeadline: string;
  }
>;

export type EarlyWithdrawalApplication = BaseApplication<
  'EARLY_WITHDRAWAL',
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export type WithdrawalApplication = BaseApplication<
  'WITHDRAWAL',
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export type FundPensionOpeningApplication = BaseApplication<
  'FUND_PENSION_OPENING' | 'FUND_PENSION_OPENING_THIRD_PILLAR',
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
    fundPensionDetails: {
      durationYears: number | 0;
      paymentsPerYear: 1 | 4 | 12;
    };
  }
>;

export type PartialWithdrawalApplication = BaseApplication<
  'PARTIAL_WITHDRAWAL',
  {
    depositAccountIBAN: string;
    cancellationDeadline: string;
    fulfillmentDate: string;
  }
>;

export type ThirdPillarWithdrawalApplication = BaseApplication<
  'WITHDRAWAL_THIRD_PILLAR',
  {
    depositAccountIBAN: string;
    cancellationDeadline: null;
    fulfillmentDate: string;
  }
>;

export type SavingsFundPaymentCancellationCommand = {
  paymentId: string;
};

export type ApplicationType =
  | 'TRANSFER'
  | 'STOP_CONTRIBUTIONS'
  | 'RESUME_CONTRIBUTIONS'
  | 'EARLY_WITHDRAWAL'
  | 'WITHDRAWAL'
  | 'PAYMENT'
  | 'SAVING_FUND_PAYMENT'
  | 'PAYMENT_RATE'
  | 'FUND_PENSION_OPENING'
  | 'FUND_PENSION_OPENING_THIRD_PILLAR'
  | 'PARTIAL_WITHDRAWAL'
  | 'WITHDRAWAL_THIRD_PILLAR';

export type ApplicationStatus = 'PENDING' | 'COMPLETE' | 'FAILED';

export interface BaseApplication<Type extends ApplicationType, Details> {
  id: number;
  status: ApplicationStatus;
  creationTime: string;

  type: Type;
  details: Details;
}

export interface Fund {
  isin: string;
  name: string;
  nav: number | null;
  pillar: Pillar;
  managementFeeRate: number;
  ongoingChargesFigure: number;
  fundManager: FundManager;
  status: FundStatus;
  inceptionDate: string;
}

export type Pillar = 2 | 3;

export type FundStatus =
  | 'ACTIVE'
  | 'LIQUIDATED'
  | 'SUSPENDED'
  | 'CONTRIBUTIONS_FORBIDDEN'
  | 'PAYOUTS_FORBIDDEN';

export interface FundBalance {
  fund: Fund;
  value: number;
  unavailableValue: number;
  currency: string;
  activeContributions: boolean;
  contributions: number;
  subtractions: number;
  profit: number;
  units: number;
}

export interface SourceFund {
  fundManager: FundManager;
  activeFund: boolean;
  name: string;
  pillar: Pillar;
  managementFeePercent: number;
  isin: string;
  price: number;
  unavailablePrice: number;
  currency: string;
  ongoingChargesFigure: number;
  contributions: number;
  subtractions: number;
  profit: number;
  units: number;
}

export interface FundManager {
  name: string;
}

export interface CancellationMandate {
  mandateId: number;
}

export interface Mandate {
  id: number;
  pillar: Pillar;
}

interface Address {
  countryCode: string;
}

interface PaymentRates {
  current: PaymentRate;
  pending: PaymentRate | null;
}

export interface User {
  id: number;
  personalCode: string;
  firstName: string;
  lastName: string;
  email: string;
  phoneNumber: string;
  memberNumber: number | null;
  memberJoinDate: string | null;
  pensionAccountNumber: string;
  secondPillarPikNumber: string;
  address: Address;
  secondPillarActive: boolean;
  thirdPillarActive: boolean;
  dateOfBirth: string;
  age: number;
  retirementAge: number;
  secondPillarPaymentRates: PaymentRates;
  secondPillarOpenDate: string | null;
  thirdPillarInitDate: string | null;
  contactDetailsLastUpdateDate: string | null;
}

export interface MemberLookup {
  id: number;
  memberNumber: number;
  firstName: string;
  lastName: string;
  personalCode: string;
}

export interface UserConversion {
  secondPillar: Conversion;
  thirdPillar: Conversion;
  weightedAverageFee: number;
}

export interface Amount {
  yearToDate: number;
  lastYear: number;
  total: number;
}

export interface Conversion {
  selectionComplete: boolean;
  selectionPartial: boolean;
  transfersComplete: boolean;
  transfersPartial: boolean;
  paymentComplete: boolean;
  pendingWithdrawal: boolean;
  contribution: Amount;
  subtraction: Amount;
  weightedAverageFee: number;
}

export interface CapitalRow {
  type: CapitalType;
  contributions: number;
  profit: number;
  value: number;
  unitCount: number;
  unitPrice: number;
  currency: Currency;
}

export interface CapitalTotal {
  unitAmount: number;
  totalValue: number;
  unitPrice: number;
}

export interface CapitalEvent {
  date: string;
  type: CapitalType;
  value: number;
  currency: Currency;
}

export type CapitalType =
  | 'CAPITAL_PAYMENT'
  | 'UNVESTED_WORK_COMPENSATION'
  | 'WORK_COMPENSATION'
  | 'MEMBERSHIP_BONUS'
  | 'CAPITAL_ACQUIRED';

export interface AmlCheck {
  type: string;
  success: boolean;
}

export interface ErrorResponse {
  body: { errors: [{ code: string }] };
}

export interface Token {
  accessToken: string;
  refreshToken: string;
}

export type LoginMethod = 'MOBILE_ID' | 'SMART_ID' | 'ID_CARD' | 'PARTNER';
export type SigningMethod = 'MOBILE_ID' | 'SMART_ID' | 'ID_CARD';
export interface AuthenticationPrincipal extends Token {
  loginMethod: LoginMethod;
  signingMethod: SigningMethod;
}

export interface MobileSignatureResponse {
  challengeCode: string | null; // Can be null during first request of Smart-ID
}

export interface MobileSignatureStatusResponse {
  statusCode: 'SIGNATURE' | 'OUTSTANDING_TRANSACTION';
  challengeCode: string;
}

export interface IdCardSignatureResponse {
  hash: string;
}

export interface IdCardSignatureStatusResponse {
  statusCode: string;
}

export interface Payment {
  type: PaymentType;
  paymentChannel: PaymentChannel;
  recipientPersonalCode: string;
  amount?: number;
  currency?: Currency;
}

export type Currency = 'EUR';

export type PaymentType = 'SINGLE' | 'RECURRING' | 'EMPLOYER' | 'GIFT' | 'MEMBER_FEE' | 'SAVINGS';

export type PaymentChannel =
  | 'SWEDBANK'
  | 'LHV'
  | 'SEB'
  | 'LUMINOR'
  | 'COOP'
  | 'COOP_WEB'
  | 'PARTNER'
  | 'TULUNDUSUHISTU';

export interface Authentication {
  authenticationHash: string;
  challengeCode: string;
}

export interface Transaction {
  amount: number;
  currency: Currency;
  time: string;
  isin: string;
  type: TransactionType;
}

export type TransactionType = 'CONTRIBUTION_CASH' | 'CONTRIBUTION_CASH_WORKPLACE' | 'SUBTRACTION';
export interface BaseContribution {
  time: string;
  sender: string;
  amount: number;
  currency: Currency;
  pillar: Pillar;
}

export interface SecondPillarContribution extends BaseContribution {
  pillar: 2;
  additionalParentalBenefit: number;
  employeeWithheldPortion: number;
  socialTaxPortion: number;
  interest: number;
}

export interface ThirdPillarContribution extends BaseContribution {
  pillar: 3;
}

export type Contribution = SecondPillarContribution | ThirdPillarContribution;

export interface MandateDeadlines {
  transferMandateFulfillmentDate: string;
  periodEnding: string;
  withdrawalFulfillmentDate: string;
  withdrawalLatestFulfillmentDate: string;
  earlyWithdrawalFulfillmentDate: string;
  transferMandateCancellationDeadline: string;
  withdrawalCancellationDeadline: string;
  earlyWithdrawalCancellationDeadline: string;
  paymentRateDeadline: string;
  paymentRateFulfillmentDate: string;
  secondPillarContributionEndDate: string;
}

export interface PaymentLink {
  url: string;
}

export type MemberCapitalListingType = 'BUY' | 'SELL';

export type MemberCapitalListingCreationRequest = {
  type: MemberCapitalListingType;
  bookValue: number;
  totalPrice: number;
  currency: Currency;
  expiryDate: string;
};

export type CreateMemberCapitalListingDto = {
  type: MemberCapitalListingType;
  bookValue: number;
  totalPrice: number;
  currency: Currency;
  expiryTime: string;
};

export type MemberCapitalListing = {
  id: number;
  type: MemberCapitalListingType;
  bookValue: number;
  totalPrice: number;
  currency: Currency;
  expiryTime: string;
  createdTime: string;
  isOwnListing: boolean;
  language: 'et' | 'en';
};

export type ContactListingOwnerDto = {
  id: number;
  addPersonalCode: boolean;
  addPhoneNumber: boolean;
};

export type ContactListingOwnerResponse = {
  id: number;
  status: string;
};
