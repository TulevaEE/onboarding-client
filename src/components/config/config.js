import config from 'react-global-configuration';

const BASE_CONFIG = {
  clientCredentialsAccessToken: undefined,
  language: 'et',
  idCardUrl: 'https://id.tuleva.ee',
  memberFee: 125,
};

const ENV_CONFIGS = {
  development: {
    ...BASE_CONFIG,
    applicationUrl: 'http://localhost:3000',
    clientCredentialsAccessToken: '6b338ba2-805c-4300-9341-b38bb4ad34a9',
  },
  production: {
    ...BASE_CONFIG,
    applicationUrl: 'https://pension.tuleva.ee',
    clientCredentialsAccessToken: '705e26c1-9316-47f2-94b8-a5c6b0dfb566',
  },
  staging: {
    ...BASE_CONFIG,
    applicationUrl: 'https://staging.tuleva.ee',
    clientCredentialsAccessToken: '6b338ba2-805c-4300-9341-b38bb4ad34a9',
    idCardUrl: 'https://id-staging.tuleva.ee',
  },
  test: {
    ...BASE_CONFIG,
    applicationUrl: 'http://localhost',
    language: 'en',
    clientCredentialsAccessToken: undefined,
  },
};

export function initializeConfiguration() {
  const env = process.env.NODE_ENV || process.env.REACT_APP_ENV;
  const selectedConfig = ENV_CONFIGS[env] || BASE_CONFIG;
  config.set(selectedConfig, { freeze: false });
}

export function updateLanguage(language) {
  const env = process.env.NODE_ENV || process.env.REACT_APP_ENV;
  const selectedConfig = { ...ENV_CONFIGS[env], language };
  config.set(selectedConfig, { freeze: false });
}
