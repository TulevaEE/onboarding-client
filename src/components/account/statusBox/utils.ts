import moment from 'moment';
import { FundPensionStatus } from '../../common/apiModels/withdrawals';
import { formatMonth } from '../../common/dateFormatter';

export const getFundPensionEndFormatted = (
  fundPension: FundPensionStatus['fundPensions'][number],
) => {
  const startDate = moment(fundPension.startDate);

  return formatMonth(startDate.add(fundPension.durationYears, 'years').toISOString());
};
