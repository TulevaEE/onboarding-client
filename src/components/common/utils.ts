import { ChangeEventHandler, HTMLProps, useState } from 'react';
import { Fund } from './apiModels';

const isTruthy = (value: unknown) => !!value;

const NOT_FOUND_ITEM_CONSTANT = {}; // using this as a secret comparison reference.

// TODO why does this exist?
export function findWhere<T>(list: T[] = [], predicate: (val: T) => boolean = isTruthy) {
  const value = list.reduce((foundItem: T, current: T) => {
    if (foundItem === NOT_FOUND_ITEM_CONSTANT && predicate(current)) {
      return current;
    }
    return foundItem;
  }, NOT_FOUND_ITEM_CONSTANT as T);
  if (value === NOT_FOUND_ITEM_CONSTANT) {
    return null;
  }
  return value;
}

export function createClamper(lowerLimit = 0, upperLimit = 10) {
  return (value: number): number => Math.max(Math.min(value, upperLimit), lowerLimit);
}

export function formatAmountForCurrency(
  amount = 0,
  fractionDigits = 2,
  options: { isSigned?: boolean } = {},
) {
  return `${formatAmountForCount(amount, fractionDigits, options)}\u00A0€`;
}

export const formatAmountForCount = (
  amount = 0,
  fractionDigits = 2,
  options: { isSigned?: boolean } = {},
) => {
  const { isSigned = false } = options;
  const sign = amount > 0 && isSigned ? '+' : '';
  let formattedAmount = amount.toFixed(fractionDigits).replace(/\B(?=(\d{3})+(?!\d))/g, '\u00A0');
  if (amount < 0) {
    formattedAmount = formattedAmount.replace('-', '−');
  }
  return `${sign}${formattedAmount}`;
};

export const getFullName = (user: { firstName: string; lastName: string }) =>
  `${user.firstName} ${user.lastName}`;

export const isActive = (fund: Fund) => fund.status === 'ACTIVE';
export const isSecondPillar = (fund: Fund) => fund.pillar === 2;
export const isThirdPillar = (fund: Fund) => fund.pillar === 3;

export const isTuleva = (fund: Fund) => (fund.fundManager || {}).name === 'Tuleva';

export type TulevaSecondPillarStockFund = 'EE3600109435';
export type TulevaSecondPillarBondFund = 'EE3600109443';
export type TulevaThirdPillarFund = 'EE3600001707';
export type TulevaFundIsin =
  | TulevaSecondPillarStockFund
  | TulevaSecondPillarBondFund
  | TulevaThirdPillarFund;
export const isTulevaIsin = (value: string): value is TulevaFundIsin =>
  ['EE3600109435', 'EE3600109443', 'EE3600001707'].includes(value);

export const useNumberInput = (
  defaultValue: number | null = null,
  isValid: (inputValue: string) => boolean = () => true,
): { inputProps: Partial<HTMLProps<HTMLInputElement>>; value: number | null } => {
  const [inputValue, setInputValue] = useState(defaultValue?.toString() ?? '');
  const [value, setValue] = useState<number | null>(defaultValue);

  const handleInputChange: ChangeEventHandler<HTMLInputElement> = (event) => {
    const changedInputValue = event.target.value;

    const formattedInputValue = changedInputValue.replace(',', '.');

    if (changedInputValue === '' || isValid(changedInputValue)) {
      setInputValue(formattedInputValue);
    }

    if (changedInputValue === '') {
      setValue(null);
      return;
    }

    if (!isValid(changedInputValue)) {
      return;
    }

    const parsedValue = Number(formattedInputValue);

    if (!Number.isNaN(parsedValue)) {
      setValue(parsedValue);
    }
  };

  return {
    inputProps: {
      value: inputValue,
      onChange: handleInputChange,
      type: 'text',
      inputMode: 'decimal',
    },
    value,
  };
};
