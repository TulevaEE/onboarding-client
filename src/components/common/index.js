import * as httpModule from './http';
import * as apiModule from './api';
import * as utilsModule from './utils';
import logoModule from './logo.svg';
import LoaderModule from './loader';
import RadioModule from './radio';
import AuthenticationLoaderModule from './authenticationLoader';
import ErrorMessageModule from './errorMessage';
import InfoTooltipModule from './infoTooltip';
import ErrorAlertModule from './errorAlert';

export const http = httpModule;
export const utils = utilsModule;
export const Loader = LoaderModule;
export const Radio = RadioModule;
export const api = apiModule;
export const logo = logoModule;
export const AuthenticationLoader = AuthenticationLoaderModule;
export const ErrorMessage = ErrorMessageModule;
export const InfoTooltip = InfoTooltipModule;
export const ErrorAlert = ErrorAlertModule;
