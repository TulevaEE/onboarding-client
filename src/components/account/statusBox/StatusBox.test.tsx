import React from 'react';
import { setupServer } from 'msw/node';
import { createMemoryHistory, History } from 'history';
import { screen } from '@testing-library/react';
import { rest } from 'msw';
import { createDefaultStore, login, renderWrapped } from '../../../test/utils';
import { initializeConfiguration } from '../../config/config';
import { StatusBox } from './StatusBox';
import { ApplicationStatus, ApplicationType } from '../../common/apiModels';

jest.unmock('retranslate');

describe('Status Box', () => {
  const server = setupServer();
  let history: History;
  let renderComponent: (ui: React.ReactElement) => void;

  function testApplication() {
    return {
      id: 123,
      type: ApplicationType.EARLY_WITHDRAWAL,
      status: ApplicationStatus.PENDING,
      creationTime: new Date('December 17, 1995 03:24:00').toISOString(),
      details: {
        withdrawalTime: new Date('January 2, 1995 03:24:00').toISOString(),
        depositAccountIBAN: 'EE123123123',
      },
    };
  }

  const props = {
    memberNumber: null,
    loading: false,
    thirdPillar: null,
    conversion: {
      secondPillar: {
        selectionComplete: false,
        transfersComplete: false,
        paymentComplete: false,
        pendingWithdrawal: false,
        subtraction: { yearToDate: 0, total: 0 },
        contribution: { yearToDate: 0, total: 0 },
      },
      thirdPillar: {
        selectionComplete: false,
        transfersComplete: false,
        paymentComplete: false,
        pendingWithdrawal: false,
        contribution: { yearToDate: 0, total: 0 },
        subtraction: { yearToDate: 0, total: 0 },
      },
    },
    secondPillarFunds: [],
    thirdPillarFunds: [],
  };

  function render() {
    history = createMemoryHistory();
    const store = createDefaultStore(history as any);
    login(store);

    const { rerender } = renderWrapped(<StatusBox {...props} />, history as any, store);
    renderComponent = rerender;
  }
  beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  beforeEach(() => {
    initializeConfiguration();
    server.use(
      rest.get('http://localhost/v1/applications', (req, res, ctx) => {
        return res(ctx.status(200), ctx.json([testApplication()]));
      }),
    );
    render();
  });

  const to2ndPillarFlow = 'Edit';
  const pay3ndPillarFlow = 'Make payment';
  const toMemberFlow = 'Sign up';

  it('renders status box title', async () => {
    expect(await screen.findByText('Your choices')).toBeInTheDocument();
  });

  it('renders 2nd pillar cta', async () => {
    expect(await screen.findByText(to2ndPillarFlow)).toBeInTheDocument();
  });

  it('always renders pay Tuleva III pillar', async () => {
    expect(await screen.findByText(pay3ndPillarFlow)).toBeInTheDocument();
  });

  it('renders join Tuleva II pillar when II pillars some in Tuleva', async () => {
    const secondPillarFunds = [
      { fundManager: { name: 'NotTuleva' }, activeFund: true, pillar: 2, name: 'Fond 1' },
      { fundManager: { name: 'Tuleva' }, activeFund: true, pillar: 2, name: 'Fond 2' },
    ];
    renderComponent(<StatusBox {...props} secondPillarFunds={secondPillarFunds} />);
    expect(await screen.findByText('Edit')).toBeInTheDocument();
  });

  it('renders become Tuleva member when not member', async () => {
    expect(await screen.findByText(toMemberFlow)).toBeInTheDocument();
  });

  it('renders pending withdrawal text when user has a pending withdrawal', async () => {
    const secondPillar = {
      selectionComplete: false,
      transfersComplete: false,
      paymentComplete: false,
      pendingWithdrawal: true,
      subtraction: { yearToDate: 0, total: 0 },
      contribution: { yearToDate: 0, total: 0 },
    };
    const conversion = { ...props.conversion, secondPillar };
    renderComponent(<StatusBox {...props} conversion={conversion} />);
    expect(await screen.findByText('You are leaving II pillar')).toBeInTheDocument();
  });

  it('renders pending withdrawal button when user has a pending withdrawal', async () => {
    const secondPillar = {
      selectionComplete: true,
      transfersComplete: true,
      paymentComplete: true,
      pendingWithdrawal: true,
      subtraction: { yearToDate: 0, total: 0 },
      contribution: { yearToDate: 0, total: 0 },
    };
    const conversion = { ...props.conversion, secondPillar };
    renderComponent(<StatusBox {...props} conversion={conversion} />);
    expect(await screen.findByText('Cancel application')).toBeInTheDocument();
  });

  it('renders pending withdrawal button when user has a pending withdrawal even when they do not have Tuleva II pillar', async () => {
    const secondPillar = {
      selectionComplete: false,
      transfersComplete: false,
      paymentComplete: false,
      pendingWithdrawal: true,
      subtraction: { yearToDate: 0, total: 0 },
      contribution: { yearToDate: 0, total: 0 },
    };
    const conversion = { ...props.conversion, secondPillar };
    renderComponent(<StatusBox {...props} conversion={conversion} />);
    expect(await screen.findByText('Cancel application')).toBeInTheDocument();
  });
});
