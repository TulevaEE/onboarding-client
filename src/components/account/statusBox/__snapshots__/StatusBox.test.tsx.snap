// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status Box renders all subcomponents when present 1`] = `
<Fragment>
  <StatusBoxTitle />
  <div
    className="card card-secondary"
  >
    <Connect(SecondPillarStatusBox) />
    <Connect(ThirdPillarStatusBox) />
    <Connect(SavingsFundStatusBox) />
    <Connect(MemberStatusBox) />
  </div>
</Fragment>
`;

exports[`Status Box shows only a loader when some props missing 1`] = `<StatusBoxLoader />`;
